### PR TITLE
Removal of python-future compatibility code

### LINKED ISSUE #389: Remove future dependency
The [Python future](https://python-future.org/) dependency was introduced as part of the migration of the source code to support Python 3, more specifically in 3e17e998190340fe789c351d716429ced6a5c846.

Initial investigation of the changes required to support Python 3.12, released in October 2023, in OMERO.py points at the `future` library as one of the current blockers. A [contribution](https://github.com/PythonCharmers/python-future/pull/626) has been opened upstream to add support for Python 3.12 but given the recent activity of this repository, it is unclear whether a `future` release will happen in the near future.

Instead, it should be possible to get rid of the `future` dependency. A secondary rationale for pursuing this approach is that OMERO.py requires Python 3+ and there is no requirement for Python 2/3 compatibility.

### PR DESCRIPTION
**Pull request by sbesson**
Fixes #389 

Primarily motivated by the support of Python 3.12 in OMERO and given the current maintenance state of the [python-future](https://python-future.org/) project, this reviews the OMERO.py source code to update all paths with either Python 2 or Python 2/3 logic to pure Python 3+.

- all imports from the `future` dependency have been reviewed and cleaned up. The commits should contain some details but capturing a few specific notes on updates that might require particular attention during the review
  * `isbytes` and `bytes_to_native_str` have been replaced by `isinstance(x, bytes)` and `x.decode('utf-8')`
  * `old_div` has been replaced by `/` for path concatenation and either `/` or `//` for mathematical divisions depending on the use case
  * `long` has been removed as it is no longer a basic type in Python 3
  * `basestring` has been replaced by `str` and/or `bytes` depending on the use case
- all imports `builtins` have been cleaned up as they should be unnecessary
- all Python 2 codepaths using the value of `sys.version_info` have been cleaned up

This branch has been minimally tested in an Ubuntu 22.04 deployment (Python 3.10) with OMERO.server and OMERO.web. There are a few initial questions:
- should the `future` dependency be fully removed from `setup.py` or kept for compatibility with the downstream projects?
- considering team testing, several downstream Python projects including OMERO.web, openmicroscopy, various OMERO CLI plugins might need a similar handling. How many do we want to be patched in a similar fashion before scheduling a round of functional testing?

Initially opening as a draft PR to allow and gather feedback. I propose to get in the nightly CI builds over the week-end to get extra coverage from the integration tests and fix additional outstanding issues and we can discuss the progress next week.

### CODE REVIEW

**Comment by snoopycrimecop** [comment]
Conflicting PR. Removed from build [OMERO-python-superbuild-push#424](https://snoopycrimecop-ci.openmicroscopy.org/jenkins/job/OMERO-python-superbuild-push/424/). See the [console output](https://snoopycrimecop-ci.openmicroscopy.org/jenkins/job/OMERO-python-superbuild-push/424/consoleText) for more details.
Possible conflicts:
  - PR #207 manics 'BlitzGateway.connect raise on error'
    - src/omero/gateway/__init__.py
  - PR #387 will-moore 'Numpy pixels type'
    - src/omero/gateway/__init__.py

--conflicts

**Comment by sbesson** [comment]
Assigning a few people to start the code review. As discussed earlier today with @jburel, @joshmoore @pwalczysko @khaledk2 and @dominikl, the next proposed steps are:
- open similar changes cleaning up `future` and the legacy Python 2 compatibility code against omero-dropbox & omero-web
- work on a set of scenarios for the testing of= OMERO.web and OMERO.py
- schedule a round of collective testing

File: src/omero/cli.py [hunk 1]
```diff
@@ -1987,7 +1971,7 @@ def response(self, client, req, loops=8, ms=500, wait=None):
             self.ctx.out("Exiting immediately")
         elif wait > 0:
             ms = wait * 1000
-            ms = old_div(ms, loops)
+            ms = ms // loops
```
Comment (jburel) [inline src/omero/cli.py:1974]: ?

**Comment by jburel** [comment]
Shouldn't block like 
```
try:
    from StringIO import StringIO
    from StringIO import StringIO as BytesIO
except ImportError:
    # Python 3
    from io import StringIO
    from io import BytesIO
```
be adjusted too?

**Comment by jburel** [comment]
``'future'`` is still listed as a dependency even if it usage is removed. Is that intended for potential extension/plugin

**Comment by jburel** [comment]
The ``python_warning.py`` module should probably be removed too.

**Comment by jburel** [comment]
What about cases in  ``omero_ext``  e.g.``py_version = sys.version_info``?

**Comment by sbesson** [comment]
Thanks for the feebdback. Pushed a few commits cleaning up additional Python 2/3 compatibility blocks.

> The python_warning.py module should probably be removed too.
> 'future' is still listed as a dependency even if its usage has been removed. Is that intended for potential extension/plugin?

In both cases, I limited the scope of this PR to non-breaking changes that would be amenable for a minor release. `python_warning.py` has been deprecated and is scheduled to be removed in the next major release of OMERO.py.

As for `future`, I can see an argument for dropping it even in a minor release as per https://semver.org/#what-should-i-do-if-i-update-my-own-dependencies-without-changing-the-public-api. Ideally, any downstream component that relies on `future` should have it declared. In practice, this will not necessarily the case and OMERO.web is a good example  which does not declare `future`. How do you want to proceed here?

**Comment by will-moore** [comment]
I was hoping to test with `python 3.12` but I guess that's not available yet?

```
$ conda create -n omero-py312 -c ome python=3.12 zeroc-ice36-python
Collecting package metadata (current_repodata.json): done
Solving environment: failed with repodata from current_repodata.json, will retry with next repodata source.
Collecting package metadata (repodata.json): done
Solving environment: \ 
Found conflicts! Looking for incompatible packages.
This can take several minutes.  Press CTRL-C to abort.
failed                                                                                                                                                             

UnsatisfiableError: The following specifications were found to be incompatible with each other:

Output in format: Requested package -> Available versions

Package python conflicts for:
zeroc-ice36-python -> python[version='>=3.6,<3.7.0a0|>=3.7,<3.8.0a0|>=3.8,<3.9.0a0|>=3.9,<3.10.0a0']
python=3.12
```

**Comment by sbesson** [comment]
@will-moore Glencoe Software is not using Conda so I cannot comment on the requirements to rebuild the `zeroc-ice36-python` recipe against a new set of Python targets.

In the spirit of the work [announced recently](https://www.glencoesoftware.com/blog/2023/12/08/ice-binaries-for-omero.html), I have built a wheel package for testing these changes in a local Python 3.12 environment. The necessary additional patches are getting ported into the repository as we speak and we are working on getting new wheel packages for Python 3.12 released for Linux  and macOS. For the latter, you might be able to give the artifacts of https://github.com/glencoesoftware/zeroc-ice-py-macos-universal2/pull/2 a try.

**Comment by chris-allan** [comment]
universal2 wheels for Ice on Python 3.12 have now been released:

 * https://github.com/glencoesoftware/zeroc-ice-py-macos-universal2/releases

**Comment by sbesson** [comment]
The last few commits consume the latest release of the pre-built Ice wheel packages and add Python 3.12 to the testing matrix. 

As indicated in https://docs.python.org/3/whatsnew/3.12.html, `setuptools` is no longer pre-installed in created virtual environments. [66ca6d9](https://github.com/ome/omero-py/pull/390/commits/66ca6d954d1271cb84efd2ffe27541748ce7c53a) updates the Tox requirements accordingly

The workflow logs have several `SyntaxWarning` related to the presence of backlash characters. Quite a few of these warnings come from the generated Python code in `omero-blitz` and more specifically the docstrings. I propose to review and update all the relevant strings in this repository and capture the rest as an issue on `omero-blitz`.

**Comment by chris-allan** [comment]
> The workflow logs have several SyntaxWarning related to the presence of backlash characters. Quite a few of these warnings come from the generated Python code in omero-blitz and more specifically the docstrings. I propose to review and update all the relevant strings in this repository and capture the rest as an issue on omero-blitz.

Similar to these, @sbesson?

https://github.com/zeroc-ice/ice/issues/1560

**Comment by sbesson** [comment]
Exactly. We have quite a few of them in the docstrings of the slice generated API classes, most of them introduced in https://github.com/ome/openmicroscopy/pull/4349.

**Comment by chris-allan** [comment]
manylinux_2_28 (not unable on RHEL 7) wheels are also now available for Python 3.12:

* https://github.com/glencoesoftware/zeroc-ice-py-linux-x86_64/releases

**Review by will-moore — APPROVED** [review]
I haven't been able to test locally with Python 3.12 but the newly added tests are green. 
Browsing the code changes looks good - can't say I've checked every line but from what I see it's 👍

**Comment by sbesson** [comment]
@pwalczysko is there a quick summary of the functional testing that could be pasted here (and cross-referenced in https://github.com/ome/omero-web/pull/531)?

**Comment by pwalczysko** [comment]
> @pwalczysko is there a quick summary of the functional testing that could be pasted here (and cross-referenced in [ome/omero-web#531](https://github.com/ome/omero-web/pull/531))?

Testing was performed through https://docs.google.com/spreadsheets/d/1zI3vlcTLp7okz_Mi2LHK40NxwCkGXUGHgDqgCd7jP_A/edit?usp=sharing and there were no major issues connected with this PR. 

The minor issues were concerning the teting setup and older parts of the codebase.

**Review by pwalczysko — APPROVED** [review]

**Pull request merged**